Don't change system role if it's already selected

dzedro opened the pull request:
- Fail: https://openqa.suse.de/tests/8010789
- Verification run:
https://openqa.suse.de/tests/overview?distri=sle&build=system_role SLE
https://openqa.opensuse.org/tests/overview?build=system_role openSUSE

dzedro commented:
The SYSTEM_ROLE default is ambiguous. It' can be anything based on with what modules is system registered. e.g.
15 SP4 now is installed with default moduels without DM gnome, so default is textmode
When Dsktop modules is added but the DESKTOP is textmode without specific PATTERNS then default will be fist option Gnome

jknphy commented:
I think that default it means that there is not selection on the screen, so it should work if you just set `SYSTEM_ROLE =default`.

foursixnine commented on tests/installation/system_role.pm:
```diff
@@ -64,7 +64,7 @@ sub assert_system_role {
         assert_screen('system-role-default-system', 180);
         my $system_role = get_var('SYSTEM_ROLE', 'default');
-        change_system_role($system_role) if (get_var('SYSTEM_ROLE') && !check_var('SYSTEM_ROLE', 'default'));
+        change_system_role($system_role) if (get_var('SYSTEM_ROLE') && !check_var('SYSTEM_ROLE', 'default') && !check_screen("system-role-$system_role-selected"));
```
Hmm this might mean that https://github.com/os-autoinst/os-autoinst/blob/a9af388073889ea5a28f3613d4b40fdfcdee265d/testapi.pm#L1415 might be broken somewhere down the line.

@okurz do you have an idea of what might have caused https://openqa.suse.de/tests/8010789#step/system_role/2?

foursixnine replied:
@dzedro Line33 is the culprit of the problem, it should first do `if !check_screen("system-role-$system_role-selected"`

dzedro replied:
Yo umean 37, but not for videomode text ? `check_var('VIDEOMODE', 'text')`

foursixnine replied:
Yeah, for the `VIDEOMODE=text` the extra check has to be added too

okurz replied:
You should be able to just use `match_has_tag` instead of `check_screen` if the needles used in the above `assert_screen` include the tag you look for. Keep in mind that needles can have multiple tags.

dzedro replied:
But they must not e.g. VR https://openqa.suse.de/tests/8013200/modules/system_role/steps/2/edit

foursixnine reviewed (approved):
I don't mind the extra check.

dzedro commented:
> 

I guess you are no using it on 15SP4+

dzedro commented:
> I think that default it means that there is not selection on the screen, so it should work if you just set `SYSTEM_ROLE =default`. The one failing has `SYSTEM_ROLE=textmode`. But I guess if you want to make it work in both ways because you need to env variable later, then your change should be fine (note: we are not using this module anymore in YaST group).

I guess you don't use this modules on 15SP4+

dzedro commented:
btw instead of `assert_screen 'system-role-default-system'` and many needles nobody care about, there could be just save_screen

foursixnine commented:
> I think that default it means that there is not selection on the screen, so it should work if you just set `SYSTEM_ROLE =default`. The one failing has `SYSTEM_ROLE=textmode`. But I guess if you want to make it work in both ways because you need to env variable later, then your change should be fine (note: we are not using this module anymore in YaST group).

See: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/14071#pullrequestreview-859463485

> The SYSTEM_ROLE default is ambiguous. It' can be anything based on with what modules is system registered. e.g. 15 SP4 now is installed with default moduels without DM gnome, so default is textmode. https://openqa.suse.de/tests/8010791#step/system_role/1

Somebody actually added SYSTEM_ROLE=textmode between the last and this build, which triggered the call to  `change_system_role` which triggered the bug that you see at: https://openqa.suse.de/tests/8010789#step/system_role/1

_Default system was selected, but a missing check screen* due to differences beetween types of installations, made openQA what it was supposed to do: keep searching for the one it was looking for_
https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/cae6400/tests/installation/system_role.pm#L31

![](https://media.giphy.com/media/4560Nv2656Gv0Lvp9F/giphy.gif)

> When Desktop module is added but the DESKTOP is textmode without specific PATTERNS then default will be fist option Gnome. https://openqa.suse.de/tests/8010805#step/system_role/1 Not sure what's the purpose of SYSTEM_ROLE _default_

Default means, default according to the selected products (https://progress.opensuse.org/issues/37504), so we should have: default-gnome, default-textmode and so on needles.

dzedro commented:
I added it, what do you mean with bug, it is not bug!
I added it to fix `create_hdd_textmode`
https://openqa.suse.de/tests/7998054 -> https://openqa.suse.de/tests/8017919
Expecting 'multi-user', but getting 'graphical' is normal. https://openqa.suse.de/tests/7998054#step/installation_overview/6
You can change system target (DM) via system role, patterns or manually, but expecting multi-user with **default** system role gnome is wrong.

jknphy commented:
> multi-user

that failure I guess it because it has DESKTOP=textmode, https://openqa.suse.de/tests/7998054#step/installation_overview/6 right? because it is a textmode installation but the installed system is graphical, installation is controlled by videmode.

dzedro commented:
> > multi-user
> 
> that failure I guess it because it has DESKTOP=textmode, https://openqa.suse.de/tests/7998054#step/installation_overview/6 right? because it is a textmode installation but the installed system is graphical, installation is controlled by videmode.

The failure happened because there is no system role or patterns which would set the textmode.
After I defined system role, another test failed because systom role test is broken.
videomode does only define if installation is ncurses or gui. https://openqa.suse.de/tests/7998054#step/bootloader/9

foursixnine reviewed (approved):
Thanks @dzedro!

dzedro commented:
Done.
Added VRs
It can happen there will be failures due to missing needles, but I will take care of that.
It can also happen removed `set_var` will cause failures, but I could not find tests where that could happen. :crossed_fingers: 
If that happens then I will add the set_var for that specific case.

btw the failure which triggered this PR can be avoided by just removing DESKTOP -> https://dzedro.suse.cz/tests/20613
The `check_default_target` and interaction between variables DESKTOP, SYSTEM_ROLE & PATTERNS do require improvement.

b10n1k commented on tests/installation/system_role.pm:
```diff
@@ -34,48 +34,37 @@ sub change_system_role {
-        assert_screen('before-role-selection', 180);
-        change_system_role(get_var('SYSTEM_ROLE', get_var('DESKTOP')));
+    assert_screen('before-role-selection', 180);
+    if (is_opensuse || (get_var('SYSTEM_ROLE') && !check_var('SYSTEM_ROLE', 'default'))) {
```
are you sure you need `get_var('SYSTEM_ROLE')`. seems like `check_var` already does this.

dzedro replied:
You miss the `!`

dzedro replied:
@b10n1k You are right, I didn't have to use `get_var`. Sorry didn't read it good enough.

b10n1k commented on tests/installation/system_role.pm:
```diff
@@ -34,48 +34,37 @@ sub change_system_role {
+    assert_screen('before-role-selection', 180);
+    if (is_opensuse || (get_var('SYSTEM_ROLE') && !check_var('SYSTEM_ROLE', 'default'))) {
+        # on opensuse is system role 1:1 with DESKTOP, on SLE change system role if it's defined
+        my $system_role = is_opensuse ? get_var('SYSTEM_ROLE', get_var('DESKTOP')) : get_var('SYSTEM_ROLE');
```
let me know if this looks more clear to you
`is_sle ? get_var('SYSTEM_ROLE') : get_var('DESKTOP');`
iiuc this is the same thing. wdyt?

dzedro replied:
It's not same.
